src/AWSMSKAuthTokenGenerator.cs

```diff
@@ -1,225 +1,306 @@
-
-            request.UseQueryString = true;
-            request.HttpMethod = HTTPMethod;
-            request.Parameters.Add(XAmzExpires, ExpiryDuration.TotalSeconds.ToString(CultureInfo.InvariantCulture));
-            request.Parameters.Add(ActionKey, ActionValue);
-            string hostName = String.Format(HostnameStringFormat, region.SystemName);
-            request.Endpoint = new UriBuilder(Scheme, hostName).Uri;
-
-            var immutableCredentials = credentials.GetCredentials();
-            if (immutableCredentials.UseToken)
-            {
-                request.Parameters[XAmzSecurityToken] = immutableCredentials.Token;
-            }
-
-            var signingResult = await Task.Run(() => AWS4PreSignedUrlSigner.SignRequest(request, null, new RequestMetrics(), immutableCredentials.AccessKey,
-                immutableCredentials.SecretKey, ServiceName, region.SystemName));
-
-            var authorization = "&" + signingResult.ForQueryParameters;
-            var url = AmazonServiceClient.ComposeUrl(request);
-
-            String userAgent = "&User-Agent=" + getUserAgent();
-            String authTokenString = url.AbsoluteUri + userAgent + authorization;
-
-            byte[] byteArray = System.Text.UTF8Encoding.UTF8.GetBytes(authTokenString);
-
-            long expiryMs = new DateTimeOffset(signingResult.DateTime.Add(ExpiryDuration)).ToUnixTimeSeconds() * 1000;
-            return (Convert.ToBase64String(byteArray).Replace('+', '-').Replace('/', '_').TrimEnd('='),  expiryMs);
+            throw new ArgumentNullException(nameof(credentials));
         }
 
-        private static String getUserAgent()
+        var immutableCredentials = useAsync ? await credentials.GetCredentialsAsync() : credentials.GetCredentials();
+
+        _logger.LogDebug("Generating auth token using credentials with access key id: {accessKey}", immutableCredentials.AccessKey);
+
+        var authTokenRequest = new GenerateMSKAuthTokenRequest();
+        IRequest request = new DefaultRequest(authTokenRequest, ServiceName);
+
+        request.UseQueryString = true;
+        request.HttpMethod = HTTPMethod;
+        request.Parameters.Add(XAmzExpires, ExpiryDuration.TotalSeconds.ToString(CultureInfo.InvariantCulture));
+        request.Parameters.Add(ActionKey, ActionValue);
+        var hostName = string.Format(HostnameStringFormat, region.SystemName);
+        request.Endpoint = new UriBuilder(Scheme, hostName).Uri;
+
+        if (immutableCredentials.UseToken)
         {
-            return "aws-msk-iam-sasl-signer-net-" + SignerVersion.CurrentVersion;
+            request.Parameters[XAmzSecurityToken] = immutableCredentials.Token;
         }
 
+        var signingResult = AWS4PreSignedUrlSigner.SignRequest(request, null, new RequestMetrics(),
+            immutableCredentials.AccessKey,
+            immutableCredentials.SecretKey, ServiceName, region.SystemName);
+
+        var authorization = signingResult.ForQueryParameters;
```

Comment:
Do we not need that `&` anymore?

Reply:
Yes, I have just removed all the logic inside the string interpolation:
`$"{url.AbsoluteUri}&{GetUserAgent()}&{authorization}"` The last & is the & that you mentioned.

Reply:
Ah, missed that it moved, thanks!